### PR TITLE
mgr/dashboard: set bucket policies

### PR DESCRIPTION
**Pull request by Pegonzal**
image

This PR:
- Adds support to set bucket policies through the Dashboard.
- Rename rgw bucket policy from 'policy' to 'bucket policy'
- Fix: hide Tags when none are present on bucket list details and sets the form dirty after deleting a tag
- Added service to manage the formatting of a textArea that works with json

Fixes: https://tracker.ceph.com/issues/63942

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by Pegonzal** [comment]
@cloudbehl @nizamial09 we talked about adding a helper here, for the policies, with a short explanation and a link to some examples. Where would be appropiate to link to?

**Comment by Pegonzal** [comment]
@cloudbehl updated the PR, added a helper with a link and fixxed linting issues

**Comment by Pegonzal** [comment]
jenkins test make check

**Comment by cloudbehl** [comment]
1. Lets call it Bucket Policy rather than policy both in listing and creation.
2. Let use info bar for displaying the link. Also use policy generator link (https://awspolicygen.s3.amazonaws.com/policygen.html) .
3. Also AWS use this link for s3 examples(https://docs.aws.amazon.com/AmazonS3/latest/userguide/example-bucket-policies.html?icmpid=docs_amazons3_console) .

**Comment by cloudbehl** [comment]
AWS show policy info like this, just for example. 

Screenshot from 2024-01-16 14-38-16

**Comment by Pegonzal** [comment]
> 1. Lets call it Bucket Policy rather than policy both in listing and creation.
> 2. Let use info bar for displaying the link. Also use policy generator link (https://awspolicygen.s3.amazonaws.com/policygen.html) .
> 3. Also AWS use this link for s3 examples(https://docs.aws.amazon.com/AmazonS3/latest/userguide/example-bucket-policies.html?icmpid=docs_amazons3_console) .

Updated the PR with your suggestions @cloudbehl!

File: src/pybind/mgr/dashboard/frontend/src/app/shared/forms/crud-form/formly-textarea-type/formly-textarea-type.component.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import { Component, ViewChild, ElementRef } from '@angular/core';
 import { FieldType, FieldTypeConfig } from '@ngx-formly/core';
+import { TextAreaJsonFormatterService } from '~/app/shared/services/text-area-json-formatter.service';
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/forms/crud-form/formly-textarea-type/formly-textarea-type.component.ts:3]: i don't see this file in the commit. probably you missed it while committing?

Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/forms/crud-form/formly-textarea-type/formly-textarea-type.component.ts:3]: You are absoluteley right!
Fixed it, pushed the missing files @nizamial09

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts [hunk 2]
```diff
@@ -70,6 +81,7 @@ export class RgwBucketFormComponent extends CdForm implements OnInit, AfterViewC
     private rgwUserService: RgwUserService,
     private notificationService: NotificationService,
     private rgwEncryptionModal: RgwBucketEncryptionModel,
+    private textAreaJsonFormatter: TextAreaJsonFormatterService,
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:84]: Let's name it properly so we know its a service. and this applies to all other parts in the code.

```suggestion
    private textAreaJsonFormatterService: TextAreaJsonFormatterService,
```

**Review by nizamial09 — APPROVED** [review]
Everything looks good to me, thank you @Pegonzal 

just one thought, what about something like this
image

and we could also include ACLs in here since it also talks about the permissions. So all can fit under one umbrella?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html [hunk 3]
```diff
@@ -410,6 +410,39 @@
...
+              Policy examples
+            </button>
+            <button type="button"
+                    id="example-generator-button"
+                    class="btn btn-light float-end my-3"
+                    (click)="openUrl('https://awspolicygen.s3.amazonaws.com/policygen.html')"
+                    i18n>
+              <i [ngClass]="[icons.externalUrl]"></i>
+              Policy generator
+            </button>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:443]: you can use a [button-group component](https://getbootstrap.com/docs/5.0/components/button-group/) to group these two together so it looks much more nicer

**Comment by Pegonzal** [comment]
> Everything looks good to me, thank you @Pegonzal
> 
> just one thought, what about something like this image
> 
> and we could also include ACLs in here since it also talks about the permissions. So all can fit under one umbrella?

I think it makes sense, I'll add the changes in this PR, I'll call the group as `Policies` and then we can have `Bucket policy` and `ACL` in it to follow current naming conventions

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts (hunk 2 above)
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts]: Got it

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html (hunk 3 above)
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html]: Oh nice, added it!

**Review by cloudbehl — APPROVED** [review]

**Comment by nizamial09** [comment]
should we be worried about this e2e failure? @Pegonzal 
```
  1) RGW buckets page
       create, edit & delete bucket tests
         should edit bucket:
     AssertionError: Timed out retrying after 120000ms: expected '<input#versioning.custom-control-input.ng-untouched.ng-pristine.ng-valid>' not to be 'checked'
      at BucketsPageHelper.edit (cypress/e2e/rgw/buckets.po.ts:101:35)
      at descriptor.value [as edit] (cypress/e2e/page-helper.po.ts:25:11)
      at Context.eval (cypress/e2e/rgw/buckets.e2e-spec.ts:26:14)
      ```

**Comment by Pegonzal** [comment]
@nizamial09  seems like it's an ongoing thing from #53021 

> should we be worried about this e2e failure? @Pegonzal
> 
> ```
>   1) RGW buckets page
>        create, edit & delete bucket tests
>          should edit bucket:
>      AssertionError: Timed out retrying after 120000ms: expected '<input#versioning.custom-control-input.ng-untouched.ng-pristine.ng-valid>' not to be 'checked'
>       at BucketsPageHelper.edit (cypress/e2e/rgw/buckets.po.ts:101:35)
>       at descriptor.value [as edit] (cypress/e2e/page-helper.po.ts:25:11)
>       at Context.eval (cypress/e2e/rgw/buckets.e2e-spec.ts:26:14)
>       ```
> ```

**Comment by nizamial09** [comment]
> @nizamial09 seems like it's an ongoing thing from #53021
> 
> > should we be worried about this e2e failure? @Pegonzal
> > ```
> >   1) RGW buckets page
> >        create, edit & delete bucket tests
> >          should edit bucket:
> >      AssertionError: Timed out retrying after 120000ms: expected '<input#versioning.custom-control-input.ng-untouched.ng-pristine.ng-valid>' not to be 'checked'
> >       at BucketsPageHelper.edit (cypress/e2e/rgw/buckets.po.ts:101:35)
> >       at descriptor.value [as edit] (cypress/e2e/page-helper.po.ts:25:11)
> >       at Context.eval (cypress/e2e/rgw/buckets.e2e-spec.ts:26:14)
> >       ```
> > ```

but that PR has all the checks passing..

**Comment by Pegonzal** [comment]
jenkins test api

**Comment by Pegonzal** [comment]
jenkins test make check

**Comment by Pegonzal** [comment]
jenkins test window

**Comment by Pegonzal** [comment]
jenkins test windows

**Pull request merged**